pkg/project/auth/watch.go

```diff
@@ -0,0 +1,217 @@
+	namespaces, _ := authCache.List(userInfo)
+	knownProjects := sets.String{}
+	for _, namespace := range namespaces.Items {
+		knownProjects.Insert(namespace.Name)
+	}
+
+	w := &userProjectWatcher{
+		username: username,
+		groups:   groups,
+
+		cacheIncoming: make(chan watch.Event, 1000),
+		cacheError:    make(chan error, 1),
+		outgoing:      make(chan watch.Event),
+		userStop:      make(chan struct{}),
+
+		projectCache:  projectCache,
+		authCache:     authCache,
+		knownProjects: knownProjects,
+	}
+	w.emit = func(e watch.Event) {
+		w.outgoing <- e
+	}
+	return w
+}
+
+func (w *userProjectWatcher) GroupMembershipChanged(namespaceName string, latestUsers, lastestGroups, removedUsers, removedGroups, addedUsers, addedGroups sets.String) {
+	removed := removedUsers.Has(w.username) || removedGroups.HasAny(w.groups...)
+	added := addedUsers.Has(w.username) || addedGroups.HasAny(w.groups...)
+	hasAccess := latestUsers.Has(w.username) || lastestGroups.HasAny(w.groups...)
+
+	switch {
+	case removed:
+		if !w.knownProjects.Has(namespaceName) {
+			return
+		}
+		w.knownProjects.Delete(namespaceName)
+
+		select {
+		case w.cacheIncoming <- watch.Event{
+			Type:   watch.Deleted,
+			Object: convertNamespace(&kapi.Namespace{ObjectMeta: kapi.ObjectMeta{Name: namespaceName}}),
+		}:
+		default:
+			// remove the watcher so that we wont' be notified again and block
+			w.authCache.RemoveWatcher(w)
+			w.cacheError <- errors.New("delete notification timeout")
+		}
+
+	case added || hasAccess:
+		namespace, err := w.projectCache.GetNamespace(namespaceName)
+		if err != nil {
+			utilruntime.HandleError(err)
+			return
+		}
+		event := watch.Event{
+			Type:   watch.Added,
+			Object: convertNamespace(namespace),
+		}
+
+		// if we already have this in our list, then we're getting notified because the object changed
+		if w.knownProjects.Has(namespaceName) {
+			event.Type = watch.Modified
+		}
+		w.knownProjects.Insert(namespace.Name)
+
+		select {
+		case w.cacheIncoming <- event:
+		default:
+			// remove the watcher so that we won't be notified again and block
+			w.authCache.RemoveWatcher(w)
+			w.cacheError <- errors.New("add notification timeout")
+		}
+
+	}
+
+}
+
+func convertNamespace(namespace *kapi.Namespace) *projectapi.Project {
+	return &projectapi.Project{
+		ObjectMeta: namespace.ObjectMeta,
+		Spec: projectapi.ProjectSpec{
+			Finalizers: namespace.Spec.Finalizers,
+		},
+		Status: projectapi.ProjectStatus{
+			Phase: namespace.Status.Phase,
+		},
+	}
+}
+
+// translate pulls stuff from etcd, converts, and pushes out the outgoing channel. Meant to be
+// called as a goroutine.
+func (w *userProjectWatcher) Watch() {
+	defer close(w.outgoing)
+	defer func() {
+		// when the watch ends, always remove the watcher from the cache to avoid leaking.
+		w.authCache.RemoveWatcher(w)
+	}()
+	defer utilruntime.HandleCrash()
+
+	for {
+		select {
+		case err := <-w.cacheError:
+			if err != nil {
```

Comment:
is this needed? do we ever send a nil err to the cacheError channel or close it?